bibtex-completion.el

```diff
@@ -478,6 +493,45 @@ for string replacement."
 (defvar bibtex-completion-cached-notes-keys nil
   "A cache storing notes keys obtained when the bibliography was last parsed.")
 
+(defun bibtex-completion--read ()
+  "Select BibTeX entries in completion system."
+  ;; define a completion function that defaults to completing-read, but can be overridden
+  (list (cdr (assoc (bibtex-completion--completing-read) (bibtex-completion--get-candidates)))))
+
+(defun bibtex-completion--completing-read ()
+  "Read bibtex-completion entries for completion using completing-read."
+  (bibtex-completion-init)
+  (completing-read
+   "BibTeX entries: "
+   (lambda (string predicate action)
+     (if (eq action 'metadata)
+         '(metadata
+           ;; (annotation-function . bibtex-completion--annotation)
+           (category . bibtex))
+       (complete-with-action action (bibtex-completion--get-candidates) string predicate)))))
+
+(defun bibtex-completion--get-candidates ()
+  "Return all keys from bibtex-completion-candidates."
+  (cl-loop
+   for candidate in (bibtex-completion-candidates)
+   collect
+   (cons
+    ;; Here we concat the string for display with an invisible string enhanced
+    ;; for search, when bibtex-completion-additional-search-fields is non-nil.
+    (concat(bibtex-completion-format-entry candidate (1- (frame-width)))
+           (propertize (bibtex-completion--get-extra-search-data candidate) 'invisible t))
+    (cdr (assoc "=key=" candidate)))))
+
+(defun bibtex-completion--get-extra-search-data (candidate)
+  "Return extended search metadata as string."
+  ;; TODO
+  (if bibtex-completion-additional-search-fields
+      ; if the data is present, pull its value(s), join into a single string
+      ;(cl-loop
+      ; for field in (bibtex-completion-additional-search-fields)
+      ;(cdr (assoc field (cdr(candidate))))
+      ))
```

Comment:
This is the last thing that needs to be resolved. It's just a simple function, but my elisp ...